api: change IP address fields over to netip types

corhere opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

- Closes #50944 
- Requires #50710 

**- What I did**
Changed all API struct fields which used `string`, `net.IP` or `net.IPNet` types to represent an IP address or CIDR prefix to instead use `netip.Addr` and `netip.Prefix`. This change has no impact on the JSON encoding of the structs. Unmarshaling a JSON object with ill-formed IP values for netip-typed fields will fail fast, simplifying downstream validation.

**- How I did it**
I audited all API structs to identify the fields that encoded IP addresses or CIDR prefixes. I elected to "boil the ocean," modifying all the daemon code to work with the new API structs, as it turned out to be pretty straightforward to do. Iterators made quick work of mapping slices and maps of values between string, netip and net types. The majority of the work was modifying tests. Given how heavily used the client is in test code, the only alternative would have been to commit a complete second copy of the api and client modules without the netip changes.

**- How to verify it**
Existing tests

**- Human readable description for the release notes**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section.

NOTE: Only fill this section if changes introduced in this PR are user-facing.
The PR must have a relevant impact/ label.
-->
```markdown changelog
- Go-SDK: IP addresses and subnets are now of type `netip.Addr` and `netip.Prefix`, respectively.
```

**- A picture of a cute animal (not mandatory but encouraged)**

corhere committed:
api/types/swarm: use netip types as appropriate

Change the types for IP address and prefix struct fields to netip.Addr
and netip.Prefix for convenience. Fields such as
swarm.InitRequest.ListenAddr which may encode non-numeric values such as
a network interface name have not been modified.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
api/types/system: use netip types where appropriate

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
api/types/network: use netip types as appropriate

And generate the ServiceInfo struct from the Swagger spec.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
api/types/container: use netip types as appropriate

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
internal: move sliceutil from daemon/internal

These utilities are very handy to use in integration tests, too. Move
the package so it can be imported by them.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
internal/sliceutil: map nil to nil

As there is a meaningful distinction between a nil slice and an empty
but non-nil slice in some contexts, having sliceutil.Map return an empty
slice when passed a nil slice can introduce subtle bugs. Modify Map to
return a nil slice when passed a nil slice.

Signed-off-by: Cory Snider <csnider@mirantis.com>

thaJeztah commented on integration/system/disk_usage_test.go:
```diff
@@ -258,7 +260,7 @@ func TestDiskUsage(t *testing.T) {
 
 					du, err := apiClient.DiskUsage(ctx, tc.options)
 					assert.NilError(t, err)
-					assert.DeepEqual(t, du, tc.expected)
+					assert.DeepEqual(t, du, tc.expected, cmpopts.EquateComparable(netip.Addr{}, netip.Prefix{}))
```
Not related to your changes, but those DiskUsage types continue giving; I recall we already reduced their scope, but I think it's still including far too much information for the purpose the response was created for;

The CLI itself discards most of the information by default;

```bash
docker system df --format json
{"Active":"9","Reclaimable":"42.88GB (95%)","Size":"44.7GB","TotalCount":"45","Type":"Images"}
{"Active":"6","Reclaimable":"311.3kB (47%)","Size":"659.5kB","TotalCount":"20","Type":"Containers"}
{"Active":"2","Reclaimable":"6.166GB (93%)","Size":"6.577GB","TotalCount":"17","Type":"Local Volumes"}
{"Active":"0","Reclaimable":"21.39GB","Size":"21.39GB","TotalCount":"466","Type":"Build Cache"}
```


And even in `--verbose` mode, most fields are discarded (output below (collapsed));


```bash
docker system df --verbose --format json | jq .Containers
```

<details>

```json
[
  {
    "Command": "\"sh -c /InstallSource/install.sh\"",
    "CreatedAt": "2025-08-25 19:46:54 +0200 CEST",
    "ID": "efacd1415cc122db6e564b024a8e087bbadaa4867fb4331e626df90b65e496d7",
    "Image": "golang",
    "Labels": "desktop.docker.io/binds/0/Source=/Users/thajeztah/Projects/cli,desktop.docker.io/binds/0/SourceKind=hostFile,desktop.docker.io/binds/0/Target=/InstallSource,desktop.docker.io/ports.scheme=v2",
    "LocalVolumes": "0",
    "Mounts": "/host_mnt/Users/thajeztah/go/src/github.com/docker/cli",
    "Names": "unruffled_benz",
    "Networks": "bridge",
    "Platform": null,
    "Ports": "",
    "RunningFor": "3 weeks ago",
    "Size": "8.19kB",
    "State": "exited",
    "Status": "Exited (126) 3 weeks ago"
  },
  {
    "Command": "\"/usr/bin/gotty --title-format 'Docker Shell - ID:{{ .Hostname }}' --permit-write nsenter1\"",
    "CreatedAt": "2025-08-22 22:17:26 +0200 CEST",
    "ID": "778b28a0d0af458de7fd71c8ed2e246c73209fcc1aafc873af2babd0ad4538bc",
    "Image": "thajeztah/ddshell",
    "Labels": "com.docker.compose.config-hash=fb3c866f20ea7cc69c1eb1055d0344cb201c277e8b5b7e2eca008542913ec2e9,com.docker.compose.container-number=1,com.docker.compose.depends_on=,com.docker.compose.image=sha256:e332ae952d6326b017e64cbb66f2514906e8d9cad56dc5d554e7cd6474f1cd45,com.docker.compose.oneoff=False,com.docker.compose.project.config_files=/Users/thajeztah/Library/Containers/com.docker.docker/Data/extensions/thajeztah_ddshell/vm/docker-compose.yml,com.docker.compose.project.working_dir=/Users/thajeztah/Library/Containers/com.docker.docker/Data/extensions/thajeztah_ddshell/vm,com.docker.compose.project=thajeztah_ddshell-desktop-extension,com.docker.compose.service=ddshell,com.docker.compose.version=2.39.2,com.docker.desktop.extension.api.version=0.1.0,com.docker.desktop.extension.name=DDShell,com.docker.desktop.extension=true,desktop.docker.io/ports.scheme=v2,desktop.docker.io/ports/8080/tcp=:8129,org.opencontainers.image.description=Docker Desktop Shell Extension,org.opencontainers.image.title=DDShell,org.opencontainers.image.vendor=Docker Inc.",
    "LocalVolumes": "0",
    "Mounts": "",
    "Names": "thajeztah_ddshell-desktop-extension-service",
    "Networks": "thajeztah_ddshell-desktop-extension_default",
    "Platform": null,
    "Ports": "0.0.0.0:8129->8080/tcp",
    "RunningFor": "3 weeks ago",
    "Size": "4.1kB",
    "State": "running",
    "Status": "Up 19 hours"
  }
]
```

</details>

thaJeztah commented on api/swagger.yaml:
```diff
@@ -2451,6 +2451,10 @@ definitions:
       VIP:
         type: "string"
         x-omitempty: false
+        x-go-type:
```
Note to self; need to create a tracking ticket to evaluate where we can add `omitempty`  (and explicit "name") for the JSON structs. Also in light of json v2, where we can be more strict.

thaJeztah commented on daemon/cluster/executor/container/container.go:
```diff
@@ -656,11 +657,11 @@ func networkCreateRequest(name string, nw *api.Network) clustertypes.NetworkCrea
-				IPRange: ic.Range,
-				Gateway: ic.Gateway,
-			})
+			var cfg network.IPAMConfig
+			cfg.Subnet, _ = netip.ParsePrefix(ic.Subnet)
+			cfg.IPRange, _ = netip.ParsePrefix(ic.Range)
+			cfg.Gateway, _ = netip.ParseAddr(ic.Gateway)
+			req.IPAM.Config = append(req.IPAM.Config, cfg)
```
Same consideration here as my other comment in the other packages; perhaps a utility would make it more clear that this is all intentional and can act as good location to describe that.

thaJeztah commented on api/types/container/nat_aliases.go:
```diff
@@ -14,11 +18,12 @@ type PortRangeProto = nat.Port
-//
-// It is currently an alias for [nat.PortBinding] but may become a concrete type in a future release.
-type PortBinding = nat.PortBinding
+type PortBinding struct {
+	// HostIP is the host IP Address
+	HostIP netip.Addr `json:"HostIp"`
+	// HostPort is the host port number
+	HostPort string
+}
```
I'm a bit lost in all the port-related types we have; so could use the extra eyes from @akerouanton and @robmry here - to verify if we don't have fields that we use / need to use as "optional" here.

thaJeztah commented on integration-cli/docker_api_containers_test.go:
```diff
@@ -507,7 +507,6 @@ func (s *DockerAPISuite) TestContainerAPIBadPort(c *testing.T) {
 		PortBindings: container.PortMap{
 			"8080/tcp": []container.PortBinding{
 				{
-					HostIP:   "",
```
(related to my previous comment, as I know we've had some cases where things were ambiguous -> "" being `0.0.0.0` or not, etc).

thaJeztah reviewed (commented):
thanks! overall looking great! ("LGTM") I left some comments, and could use the extra eyes 🤗

robmry reviewed (commented):
Excellent, this will make things much better!

I'm still working through the changes, but figured this might be worth noting now (before it gets merged if I'm too slow!) ... handling of IPv4-mapped IPv6 addresses from the API/CLI is changed by this PR. We currently error in some cases that now do something odd, and mapped addresses worked in some cases where they now fail.

(Supplying mapped addresses is an odd thing to do, but people may be doing it in cases where it works, https://github.com/moby/moby/issues/45712.)

I think we'll need a sprinkling of `Unmap()` calls, both for addresses supplied as input, and where we've used `netip.AddrFromSlice`. And, maybe some `Masked()` calls for `netip.Prefix`.

For example ...

**_Before:_**
```
# docker network create --subnet ::ffff:192.168.0.0/24 n1
Error response from daemon: failed to allocate gateway (): cannot find address pool for poolID:::/24/invalid Prefix
```

**_After:_** the IPv4 subnet is treated as IPv6 in the network config, and it's silently ignored in the container:
```
# docker network create --subnet ::ffff:192.168.0.0/24 n1
# docker inspect n1 -f '{{ json .IPAM.Config }}'
[{"Subnet":"::ffff:192.168.0.0/24"},{"Subnet":"172.19.0.0/16","Gateway":"172.19.0.1"}]
# docker run --rm -ti --network n1 busybox ip a show eth0
11: eth0@if27: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue
    link/ether e2:70:2b:b6:0c:33 brd ff:ff:ff:ff:ff:ff
    inet 172.19.0.2/16 brd 172.19.255.255 scope global eth0
       valid_lft forever preferred_lft forever
```

**_Before:_**
```
# docker network create --subnet 192.168.0.0/24 n1
# docker run --rm -ti --ip ::ffff:192.168.0.123 --network n1 --name c1 busybox ip a show eth0
11: eth0@if24: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue
    link/ether 2a:bd:d7:d0:56:78 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.123/24 brd 192.168.0.255 scope global eth0
       valid_lft forever preferred_lft forever
```

**_After:_**
```
# docker network create --subnet 192.168.0.0/24 n1
# docker run --rm -ti --ip ::ffff:192.168.0.123 --network n1 --name c1 busybox ip a show eth0
docker: Error response from daemon: invalid config for network n1: invalid endpoint settings:
* invalid IPv4 address: ::ffff:192.168.0.123
* no configured subnet or ip-range contain the IP address ::ffff:192.168.0.123
```

corhere commented on internal/netipstringer/stringer.go:
```diff
@@ -0,0 +1,25 @@
+	if !addr.IsValid() {
+		return ""
+	}
+	return addr.String()
```
No, since these are supposed to be exactly equivalent to `(netip.Addr).String()` in all cases where the value is valid. If the caller wants to string an unmapped address it should unmap the address it passes in. And there is no `(netip.Prefix).Unmap()`

corhere committed:
daemon/internal: move netiputil from libnetwork

These utilities are going to be needed elsewhere in the daemon to handle
netip values from API requests.

Signed-off-by: Cory Snider <csnider@mirantis.com>

robmry reviewed (commented):
LGTM - apart from the tests that need updates, and nits/questions ...

corhere committed:
api/t/network: validate EndpointIPAMConfig in daemon

Clients should not make assumptions about the validity of an API struct
as the set of well-formed values may differ across daemon versions.
Remove it from the API module so client-application authors are not
tempted to apply it, which would restrict the forward compatibility of
the client.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
api/t/network: move IPAM validation to daemon

Clients should not make assumptions about the validity of an API struct
as the set of well-formed values may differ across daemon versions.
Remove it from the API module so client-application authors are not
tempted to apply it, which would restrict the forward compatibility of
the client.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere committed:
api/types/registry: use netip types as appropriate

Signed-off-by: Cory Snider <csnider@mirantis.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on daemon/pkg/registry/search_test.go:
```diff
@@ -420,18 +420,6 @@ func TestSearch(t *testing.T) {
 func TestNewIndexInfo(t *testing.T) {
 	overrideLookupIP(t)
```
Ah, yes; I was right on he stubbing out / mocking; https://github.com/moby/moby/blob/b26972f9f2d69d5064642b248d72c1696cbd8317/daemon/pkg/registry/registry_test.go#L12-L33

thaJeztah replied:
These are the three that are failing; https://github.com/moby/moby/blob/b26972f9f2d69d5064642b248d72c1696cbd8317/daemon/pkg/registry/registry_test.go#L94-L96

The other two are here; common scenario looks to be matching against, e.g.`42.42.0.0/16`;

https://github.com/moby/moby/blob/b26972f9f2d69d5064642b248d72c1696cbd8317/daemon/pkg/registry/search_test.go#L633-L637

thaJeztah commented:
Oh! Welp; I think I pushed and you as well, @corhere, or is it github showing wrong? 🙈

thaJeztah reviewed (approved):
and it's all green ✅ ✅ ✅ ✅ 


This PR already had quite some eyes on it; let's bring it in!


LGTM
